lib/SILOptimizer/Mandatory/AddressLowering.cpp

```diff
@@ -596,944 +1198,1991 @@ class AddressMaterialization {
   AddressMaterialization(AddressLoweringState &pass, SILBuilder &B)
       : pass(pass), B(B) {}
 
-  SILValue initializeOperandMem(Operand *operand);
+  /// Return the address of the storage for `origValue`. This may involve
+  /// materializing projections. Record the materialized address as storage for
+  /// origValue. Called once at the definition of \p origValue.
+  SILValue materializeAddress(SILValue origValue) {
+    ValueStorage &storage = pass.valueStorageMap.getStorage(origValue);
+    if (storage.storageAddress)
+      return storage.storageAddress;
+
+    if (storage.isUseProjection) {
+      materializeUseProjectionStorage(storage, /*intoPhiOperand*/ false);
+    } else {
+      assert(storage.isDefProjection);
+      storage.storageAddress = materializeDefProjection(origValue);
+    }
+    return storage.storageAddress;
+  }
+
+  void initializeOperand(Operand *operand);
 
-  SILValue materializeAddress(SILValue origValue);
+  SILValue materializeUseProjectionStorage(ValueStorage &storage,
+                                           bool intoPhiOperand);
+
+  SILValue materializeDefProjection(SILValue origValue);
 
 protected:
-  SILValue materializeProjection(Operand *operand);
+  SILValue materializeStructExtract(SILInstruction *extractInst,
+                                    SILValue elementValue, unsigned fieldIdx);
+
+  SILValue materializeTupleExtract(SILInstruction *extractInst,
+                                   SILValue elementValue, unsigned fieldIdx);
+
+  SILValue materializeProjectionIntoUse(Operand *operand, bool intoPhiOperand);
```

Comment:
`materializeProjectionIntoUse` sounds like the same as `materializeUseProjectionStorage`.
IIUC, `materializeUseProjectionStorage` adds some caching around `materializeProjectionIntoUse`.
Maybe you can find better names for these functions.

Reply:
> IIUC, materializeUseProjectionStorage adds some caching around materializeProjectionIntoUse

These are mutually recursive for materializing use projections. I renamed the more general entry point to `recursivelyMaterializeStorage`